linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala

```diff
@@ -0,0 +1,142 @@
+
+    val FooClass = ClassName("Foo")
+
+    val jsMethodName = StringLiteral("foo")
+    val staticMethodName = m("value", Nil, IntRef)
+
+    def classDefs(pre: Boolean) = Seq(
+        mainTestClassDef(
+            consoleLog(JSMethodApply(LoadModule(FooClass), jsMethodName, Nil))
+        ),
+        classDef(
+            FooClass,
+            kind = ClassKind.ModuleClass,
+            superClass = Some(ObjectClass),
+            memberDefs = List(
+                trivialCtor(FooClass),
+                JSMethodDef(EMF, jsMethodName, Nil,
+                    if (pre) int(5)
+                    else ApplyStatic(EAF, FooClass, staticMethodName, Nil)(IntType)
+                )(EOH, None),
+                MethodDef(EMF.withNamespace(MemberNamespace.PublicStatic),
+                    staticMethodName, NON, Nil, IntType, Some(int(6)))(EOH, None)
+            )
+        )
+    )
+
+    testIncremental(classDefs(_), _ => MainTestModuleInitializers)
+  }
+
+}
+
+object IncrementalTest {
+
+  def testIncremental(
+      classDefs: Boolean => Seq[ClassDef],
+      moduleInitializers: Boolean => List[ModuleInitializer],
+      config: StandardConfig = StandardConfig())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+    Future.sequence(Seq(
+        testIncrementalStep(backward = false, classDefs, moduleInitializers, config),
+        testIncrementalStep(backward = true, classDefs, moduleInitializers, config)
+    )).map(_ => ())
+  }
+
+  private def testIncrementalStep(
+      backward: Boolean,
+      classDefs: Boolean => Seq[ClassDef],
+      moduleInitializers: Boolean => List[ModuleInitializer],
+      config: StandardConfig = StandardConfig())(
+      implicit ec: ExecutionContext): Future[Unit] = {
+
+    val outputInc = MemOutputDirectory()
+    val outputBatch = MemOutputDirectory()
+
+    val linkerInc = StandardImpl.linker(config)
+    val linkerBatch = StandardImpl.linker(config)
+
+    val logger = new ScalaConsoleLogger(Level.Error)
+
+    for {
+      minilib <- TestIRRepo.minilib
+      classDefs0 = minilib ++ classDefs(!backward).map(MemClassDefIRFile(_))
+      classDefs1 = minilib ++ classDefs(backward).map(MemClassDefIRFile(_))
+      inits0 = moduleInitializers(backward)
+      inits1 = moduleInitializers(!backward)
+      _ <- linkerInc.link(classDefs0, inits0, outputInc, logger)
+      reportInc <- linkerInc.link(classDefs1, inits1, outputInc, logger)
+      reportBatch <- linkerBatch.link(classDefs1, inits1, outputBatch, logger)
+    } yield {
+      assertModulesEqual(reportInc.publicModules, reportBatch.publicModules)
+      assertOutputEquals(outputInc, outputBatch)
+    }
+  }
+
+  private def assertModulesEqual(inc: Iterable[Report.Module],
+      batch: Iterable[Report.Module]): Unit = {
+    // Poor man's equality based on toString()
+
+    def strs(ms: Iterable[Report.Module]) =
+      ms.map(m => m.moduleID -> m.toString()).toMap
+
+    assertEquals(strs(inc), strs(batch))
```

Comment:
Consider including the value of `backwards` in the associated `message` to `assertEquals`. Since this will be called twice with the same stack trace, it's easier to figure out where's the problem if the message says in which invocation we are.